core-processor/src/processor.rs

```diff
@@ -251,82 +252,56 @@ pub fn process_executable<A: ProcessorExt + EnvExt + IntoExtInfo + 'static, E: E
     ) {
         Ok(res) => match res.kind {
             DispatchResultKind::Trap(reason) => {
-                process_error(res.dispatch, initial_nonce, res.gas_amount.burned(), reason)
+                process_error(res.dispatch, program_id, res.gas_amount.burned(), reason)
             }
             DispatchResultKind::Success => process_success(Success, res),
             DispatchResultKind::Wait => process_success(Wait, res),
             DispatchResultKind::Exit(value_destination) => {
                 process_success(Exit(value_destination), res)
             }
         },
-        Err(e) => process_error(
-            dispatch,
-            initial_nonce,
-            e.gas_amount.burned(),
-            Some(e.reason),
-        ),
+        Err(e) => process_error(dispatch, program_id, e.gas_amount.burned(), Some(e.reason)),
     }
 }
 
 /// Helper function for journal creation in message no execution case
-fn process_non_executable(dispatch: Dispatch, exit_code: ExitCode) -> Vec<JournalNote> {
+fn process_non_executable(
+    dispatch: IncomingDispatch,
+    program_id: ProgramId,
+    exit_code: ExitCode,
+) -> Vec<JournalNote> {
     // Number of notes is predetermined
     let mut journal = Vec::with_capacity(4);
 
-    let Dispatch { message, .. } = dispatch;
-
-    let message_id = message.id();
-    let value = message.value();
+    let message_id = dispatch.id();
+    let value = dispatch.value();
 
     if value != 0 {
         // Send value back
         journal.push(JournalNote::SendValue {
-            from: message.source(),
+            from: dispatch.source(),
             to: None,
             value,
         });
     }
 
     // Reply back to the message `source`
-    let reply_message = Message::new_reply(
-        crate::id::next_system_reply_message_id(message.dest(), message_id),
-        message.dest(),
-        message.source(),
-        Default::default(),
-        // Error reply value must be 0!
-        0,
-        message_id,
-        exit_code,
-    );
-    journal.push(JournalNote::SendDispatch {
-        message_id,
-        dispatch: Dispatch::new_reply(reply_message),
-    });
+    if !dispatch.is_reply() || dispatch.exit_code().expect("Checked before") == 0 {
+        let id = MessageId::generate_reply(dispatch.id(), exit_code);
```

Comment:
Here the function argument `exit_code` is used though `expect` is called on the `dispatch.exit_code()`. Do we really need to check `dispatch.exit_code()`?